src/extension.ts

```diff
@@ -349,20 +352,46 @@ class LineDecorationsGroup {
 	missedLineDecorations: vscode.DecorationOptions[] = [];
 };
 
-function createDecorationsForFile(linesDataOfFile: GcovLineData[]): LineDecorationsGroup {
-	const decorations = new LineDecorationsGroup();
+function calculateCoverageRate(linesDataOfFile: GcovLineData[]): Map<string, number> {
+	const funcLines = groupData(linesDataOfFile, x => x.function_name);
+	const funcCoverages: Map<string, number> = new Map();
 
+	for (const [funcName, lineDatas] of funcLines.entries()) {
+		let totalLineNum = 0;
+		let calledLineNum = 0;
+		for (const lineData of lineDatas.values()) {
+			totalLineNum++;
+			calledLineNum += (lineData.count > 0 ? 1 : 0);
+		}
+		funcCoverages.set(funcName, 100 * calledLineNum / totalLineNum);
+	}
+	return funcCoverages;
+}
+
+function isFirstLineOfFunction(funcsDataOfFile: GcovFunctionData[], funcName: string, lineNumber: number): boolean {
+	const funcData = funcsDataOfFile.find(x => x.name == funcName);
```

Comment:
This has quadratic runtime complexity afaik, because you are iterating over all function names for every line. This can be quite bad in large files (as the ones I'm working with quite often). Better create a map from function name to line number and do lookups in this map.